### PR TITLE
[ESI] Move services over to bundles

### PR DESCRIPTION
**Pull request by teqdruid**
Services now only support bundle ports. Bundles pretty significantly simplify services at the expense of some nomenclature complexity.

### CODE REVIEW

File: lib/Dialect/ESI/ESIServices.cpp [hunk 1]
```diff
@@ -68,33 +69,42 @@ static LogicalResult instantiateCosimEndpointOps(ServiceImplementReqOp implReq,
...
+        .replaceAllUsesWith(pack.getBundle());
 
-    auto cosim =
-        b.create<CosimEndpointOp>(loc, toClientType, clk, rst, toServerValue,
+    size_t chanIdx = 0;
+    for (BundledChannel ch : bundleType.getChannels())
+      if (ch.direction == ChannelDirection::from)
+        b.create<CosimEndpointOp>(loc, i1ch, clk, rst,
+                                  pack.getFromChannels()[chanIdx++],
                                   toStringAttr(clientNamePathAttr));
```
Comment (mortbopet) [inline lib/Dialect/ESI/ESIServices.cpp:107]: really not a big fan of unbracketed for loops when the inner nested statement breaks over multiple lines. [technically not against the style guidelines](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements) although one can argue whether this counts as "more than two levels of nesting".

File: lib/Dialect/ESI/ESIServices.cpp [hunk 2]
```diff
@@ -191,18 +205,21 @@ instantiateSystemVerilogMemory(ServiceImplementReqOp implReq,
...
       // Construct the response channel.
       auto dataValid = bb.get(i1);
       auto data = bb.get(ramDecl.getInnerType());
-      toClientResp = b.create<WrapValidReadyOp>(data, dataValid);
+      auto dataChannel = b.create<WrapValidReadyOp>(data, dataValid);
+
+      auto pack =
+          b.create<PackBundleOp>(implReq.getLoc(), req.getToClient().getType(),
+                                 dataChannel.getChanOutput());
+      Value toServer = pack.getFromChannels()[0];
```
Comment (mortbopet) [inline lib/Dialect/ESI/ESIServices.cpp:217]: random constant (`[0]`).

File: lib/Dialect/ESI/ESIServices.cpp [hunk 3]
```diff
@@ -480,73 +545,24 @@ ESIConnectServicesPass::surfaceReqs(hw::HWMutableModuleLike mod,
...
-         llvm::zip(newInst->getResults(), inst->getResults()))
-      oldV.replaceAllUsesWith(newV);
-
-    // Clone the to_server requests and wire them up to the new instance.
-    outputCounter = origNumOutputs;
-    for (Backedge newResult : newResultBackedges)
-      newResult.setValue(newInst->getResult(outputCounter++));
+    inst->insertOperands(inst->getNumOperands(), newOperands);
+    // Set the names, if we know how.
+    if (auto hwInst = dyn_cast<hw::InstanceOp>(*inst))
```
Comment (mortbopet) [inline lib/Dialect/ESI/ESIServices.cpp:562]: I suggest opening an issue to track locations where we really should be using a `HWMutableInstanceLike` interface.

File: lib/Dialect/ESI/ESIStdServices.cpp [hunk 4]
```diff
@@ -21,18 +21,24 @@
...
+                                     StringRef reqName, Type reqType,
+                                     StringRef respName, Type respType) {
+  assert(reqType || respType);
+  auto *ctxt = reqType ? reqType.getContext() : respType.getContext();
+  auto bundle = ChannelBundleType::get(
+      ctxt,
+      {BundledChannel{StringAttr::get(ctxt, reqName), ChannelDirection::to,
+                      ChannelType::get(ctxt, reqType)},
+       BundledChannel{StringAttr::get(ctxt, respName), ChannelDirection::from,
+                      ChannelType::get(ctxt, respType)}},
```
Comment (mortbopet) [inline lib/Dialect/ESI/ESIStdServices.cpp:35]: As i read it, it's acceptable behaviour that (xor) one of these channels may have a null type used in the `ChannelType`?

File: lib/Dialect/ESI/ESITypes.cpp [hunk 5]
```diff
@@ -164,6 +164,17 @@ inline ::llvm::raw_ostream &operator<<(::llvm::raw_ostream &p,
 }
 } // namespace llvm
 
+ChannelBundleType ChannelBundleType::getReversed() const {
+  SmallVector<BundledChannel, 4> reversed;
+  for (auto channel : getChannels()) {
+    ChannelDirection rev = channel.direction == ChannelDirection::from
+                               ? ChannelDirection::to
+                               : ChannelDirection::from;
```
Comment (mortbopet) [inline lib/Dialect/ESI/ESITypes.cpp:172]: mega-nit: extract this to a `static ChannelDirection flip(ChannelDirection)` function.

File: lib/Dialect/Handshake/Transforms/LowerExtmemToHW.cpp [hunk 6]
```diff
@@ -205,24 +207,22 @@ LogicalResult HandshakeLowerExtmemToHWPass::wrapESI(
...
     for (unsigned i = 0; i < memType.loadPorts; ++i) {
-      auto loadReq = b.create<esi::RequestInOutChannelOp>(
-          loc, handshake::esiWrapper(dataType), loadServicePort,
-          backedges[resIdx], b.getArrayAttr({}));
-      instanceArgsFromThisMem.push_back(loadReq);
+      auto reqPack = b.create<esi::PackBundleOp>(loc, readPortInfo.type,
+                                                 (Value)backedges[resIdx]);
+      b.create<esi::RequestToServerConnectionOp>(
+          loc, readPortInfo.port, reqPack.getBundle(), b.getArrayAttr({}));
+      instanceArgsFromThisMem.push_back(reqPack.getFromChannels()[0]);
```
Comment (mortbopet) [inline lib/Dialect/Handshake/Transforms/LowerExtmemToHW.cpp:215]: again, really not a big fan of random array indexing.

**Review by mortbopet — APPROVED** [review]
LGTM barring a few comments

File: include/circt/Dialect/ESI/ESIInterfaces.td [hunk 7]
```diff
@@ -59,77 +59,19 @@ def ServiceDeclOpInterface : OpInterface<"ServiceDeclOpInterface"> {
...
-      /*methodBody=*/"",
-      /*defaultImplementation=*/[{
-        return {};
-      }]>,
-    InterfaceMethod<
-      "Set the value headed to the server. Can assert().",
-      "void", "setToServer", (ins "Value":$newValue),
+      "Get info on a particular port.",
+      "FailureOr<ServicePortInfo>", "getPortInfo",
+      (ins "StringAttr":$portName),
```
[the remark this answers is not in the document]
Comment (teqdruid) [reply include/circt/Dialect/ESI/ESIInterfaces.td:64]: Going from something non-attr would require constructing an attr, which is not necessary given a common use case involves passing in a stringattr.

File: lib/Dialect/ESI/ESIStdServices.cpp (hunk 4 above)
Comment (teqdruid) [reply lib/Dialect/ESI/ESIStdServices.cpp:35]: Nope. There's a typo in the assert.

File: lib/Dialect/ESI/ESIServices.cpp (hunk 3 above)
Comment (teqdruid) [reply lib/Dialect/ESI/ESIServices.cpp:562]: #6319

**Comment by seldridge** [comment]
@teqdruid: I'm seeing failures with this patch:

```
RUN: at line 1: /Users/schuylereldridge/repos/github.com/llvm/circt/build/bin/circt-opt --esi-connect-services --canonicalize /Users/schuylereldridge/repos/github.com/llvm/circt/test/Dialect/ESI/services.mlir | /Users/schuylereldridge/repos/github.com/llvm/circt/build/bin/circt-opt | /Users/schuylereldridge/repos/github.com/llvm/circt/build/bin/FileCheck /Users/schuylereldridge/repos/github.com/llvm/circt/test/Dialect/ESI/services.mlir --check-prefix=CONN
+ /Users/schuylereldridge/repos/github.com/llvm/circt/build/bin/circt-opt --esi-connect-services --canonicalize /Users/schuylereldridge/repos/github.com/llvm/circt/test/Dialect/ESI/services.mlir
+ /Users/schuylereldridge/repos/github.com/llvm/circt/build/bin/circt-opt
+ /Users/schuylereldridge/repos/github.com/llvm/circt/build/bin/FileCheck /Users/schuylereldridge/repos/github.com/llvm/circt/test/Dialect/ESI/services.mlir --check-prefix=CONN
PLEASE submit a bug report to https://github.com/llvm/circt and include the crash backtrace.
Stack dump:
0.	Program arguments: /Users/schuylereldridge/repos/github.com/llvm/circt/build/bin/circt-opt --esi-connect-services --canonicalize /Users/schuylereldridge/repos/github.com/llvm/circt/test/Dialect/ESI/services.mlir
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  libLLVMSupport.dylib     0x000000010b7789c4 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  libLLVMSupport.dylib     0x000000010b776f20 llvm::sys::RunSignalHandlers() + 112
2  libLLVMSupport.dylib     0x000000010b77903c SignalHandler(int) + 304
3  libsystem_platform.dylib 0x00000001a866aa24 _sigtramp + 56
4  libCIRCTESI.dylib        0x0000000106d1aee8 void llvm::function_ref<void (mlir::Operation*)>::callback_fn<(anonymous namespace)::ESIConnectServicesPass::runOnOperation()::$_2>(long, mlir::Operation*) + 352
5  libCIRCTESI.dylib        0x0000000106d1aee8 void llvm::function_ref<void (mlir::Operation*)>::callback_fn<(anonymous namespace)::ESIConnectServicesPass::runOnOperation()::$_2>(long, mlir::Operation*) + 352
6  libCIRCTESI.dylib        0x0000000106d1ad24 void mlir::detail::walk<mlir::ForwardIterator>(mlir::Operation*, llvm::function_ref<void (mlir::Operation*)>, mlir::WalkOrder) + 180
7  libCIRCTESI.dylib        0x0000000106d1ad24 void mlir::detail::walk<mlir::ForwardIterator>(mlir::Operation*, llvm::function_ref<void (mlir::Operation*)>, mlir::WalkOrder) + 180
8  libCIRCTESI.dylib        0x0000000106d18960 (anonymous namespace)::ESIConnectServicesPass::runOnOperation() + 140
9  libMLIRPass.dylib        0x0000000106bf2038 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 580
10 libMLIRPass.dylib        0x0000000106bf2640 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 320
11 libMLIRPass.dylib        0x0000000106bf4378 mlir::PassManager::run(mlir::Operation*) + 1020
12 libMLIROptLib.dylib      0x00000001020796d0 performActions(llvm::raw_ostream&, std::__1::shared_ptr<llvm::SourceMgr> const&, mlir::MLIRContext*, mlir::MlirOptMainConfig const&) + 2692
13 libMLIROptLib.dylib      0x0000000102078bc0 mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&)::$_1>(long, std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) + 460
14 libMLIRSupport.dylib     0x00000001089d3e08 mlir::splitAndProcessBuffer(std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<mlir::LogicalResult (std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, bool, bool) + 640
15 libMLIROptLib.dylib      0x00000001020737f4 mlir::MlirOptMain(llvm::raw_ostream&, std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&) + 340
16 libMLIROptLib.dylib      0x0000000102073c60 mlir::MlirOptMain(int, char**, llvm::StringRef, mlir::DialectRegistry&) + 1068
17 circt-opt                0x00000001002d958c main + 1512
18 dyld                     0x00000001a82e3f28 start + 2236
/Users/schuylereldridge/repos/github.com/llvm/circt/test/Dialect/ESI/services.mlir:14:16: error: CONN-LABEL: expected string not found in input
```

**Comment by seldridge** [comment]
☝️ Fixed for me with https://github.com/llvm/circt/commit/3a2151c0f064125c7777e1b06fba869d887bda1a. Please post-commit review that if I'm doing something wrong. 😅

**Pull request merged**